### PR TITLE
Qmaps 1772 step by step

### PR DESCRIPTION
**Pull request by xem**
## Description
Redesign mobile step-by-step mode for itineraries
- return button on top left
- scrollable list of cards at the bottom
- scroll-snap + custom style for current step (highlighted) and past steps (grey)
- no geolocation button (for now)
- cards have a fixed height by default (2 lines of text)
- if a third line of text is present; it fades to white, calling the user to click on the card to see more
- when clicking a card, all the cards expand to show all the text they contain and the 
 map's scale moves up
- when clicking again, they all get back to the original size, and the map's scale moves down
- if a card is extremely high (> 30vh), its content becomes scrollable
- further redesign will happen later!

## Screenshots
sbs2
sbs3

### CODE REVIEW

**Comment by xem** [comment]
à discuter: il faudra peut-être retoucher les paddings du flyTo quand la map se déplace de step en step, j'ai l'impression que parfois la card cache une partie du segment de route en cours (assez rare)

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 1]
```diff
@@ -2,62 +2,79 @@ import React from 'react';
...
   static propTypes = {
     steps: PropTypes.array.isRequired,
     onClose: PropTypes.func.isRequired,
-  }
+  };
 
-  state = {
-    currentStep: 0,
-  }
+  state = { };
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:12]: You can remove that if you don't need a state…

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 2]
```diff
@@ -2,62 +2,79 @@ import React from 'react';
...
 
   componentDidMount() {
-    this.zoomToCurrentStep();
-    fire('move_mobile_geolocation_button', 65);
+    this.steps = document.querySelector('.mobile-roadmap-preview-steps');
+    this.stepList = document.querySelectorAll('.itinerary_mobile_step');
+    this.currentStep = 0;
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:17]: … but actually, this is typically something that should be stored in the state :)
This perfectly matches the definition of it: an internal value storing a part of the current component state which, when it changes, should result in a visual change.

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 3]
```diff
@@ -2,62 +2,79 @@ import React from 'react';
...
+    fire('move_mobile_bottom_ui', 105);
+    this.highlightCurrentStep();
   }
 
-  zoomToCurrentStep = () => {
-    fire('zoom_step', this.props.steps[this.state.currentStep]);
-  }
+  highlightCurrentStep = () => {
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:22]: This follows my previous comment: most of what you do imperatively in this method is the job of React.
By managing `currentStep` in the state, each change to it will trigger a re-render. In the `render` method, you'll have all the necessary pieces to assign the right `className` to each step DOM element while you render them, without the need for an external loop like this.

The main thing to keep in mind: in React, we don't render once then manipulate the DOM. Instead, we manipulate data state and the DOM always follows by reflecting changes.

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 4]
```diff
@@ -2,62 +2,79 @@ import React from 'react';
...
+    }
+  };
 
-    return <React.Fragment>
-      <div className="itinerary_mobile_step">
-        <CloseButton onClick={this.props.onClose} />
-        <RoadMapStep step={currentStep} />
-      </div>
+  toggleSize = () => {
+    this.steps.classList.toggle('expanded');
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:56]: Same for this `expanded` status, this looks like a state value.

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 5]
```diff
@@ -2,62 +2,79 @@ import React from 'react';
...
-          onClick={this.previous}
-        />
-        <button
-          className="itinerary_mobile_step_button icon-chevron-right"
-          disabled={this.state.currentStep === this.props.steps.length - 1}
-          onClick={this.next}
-        />
+  render() {
+    return <div className="itinerary_mobile_step_by_step">
+      <div className="mobile-roadmap-preview-close" onClick={this.props.onClose}>
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:62]: This could probably use the `<Button>` component, which creates a real `button` element.

Comment (bbecquet) [reply src/panel/direction/MobileRoadMapPreview.jsx:62]: I know it's consistent with the design, but this back button looks awfully small. Was it discussed?
Anyway, if not it's ok to move on with this size and come back on it later.

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 6]
```diff
@@ -2,62 +2,79 @@ import React from 'react';
...
+        onScroll={this.scroll}
+        onClick={this.toggleSize}
+      >
+        {
+          this.props.steps.map((step, index) =>
+            <div key={index} className="itinerary_mobile_step">
+              <RoadMapStep step={step}/>
+            </div>)
+        }
+        <div className="itinerary_mobile_step spacer"></div>
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:76]: Isn't it something that could be added as an `:after` pseudo element?

File: src/panel/direction/RoadMapItem.jsx [hunk 7]
```diff
@@ -7,7 +7,7 @@ const RoadMapItem = ({ children, icon, distance, className, line, ...rest }) =>
   <>
     <div className={classnames('itinerary_roadmap_item', className)} {...rest}>
       {line}
-      <Flex alignItems="center">
+      <Flex alignItems="flex-start">
```
Comment (bbecquet) [inline src/panel/direction/RoadMapItem.jsx:10]: This component is also used in the roadmaps displayed as list, so it changes the alignement of the icon in it too.

File: src/scss/includes/direction.scss [hunk 8]
```diff
@@ -287,12 +287,11 @@ button.direction_shortcut {
 
   .itinerary_mobile_step {
```
Comment (bbecquet) [inline src/scss/includes/direction.scss:288]: This is the same class you overwrite below in a more specific selector. A lot of the values declared here are overwritten (ex: `position: fixed`) or become obsolete (ex : bottom/left/right).
I suggest you merge both declaration and maybe simplifying the nesting of selectors.

File: src/scss/includes/direction.scss [hunk 9]
```diff
@@ -395,8 +368,115 @@ button.direction_shortcut {
...
+          background-color: $grey-lighter;
+
+          .itinerary_roadmap_instruction:after {
+            background: linear-gradient(#e8e8ed77, $grey-lighter);
+          }
+        }
+
+        &.active {
+          border: 1px solid $action-blue-dark;
+          box-shadow: 0 0 0 4px rgba(128, 172, 255, 0.2);
```
Comment (bbecquet) [inline src/scss/includes/direction.scss:441]: The shadow effect gets cut at the top. Probably a problem with the height of the list itself.
Capture d’écran de 2020-12-03 10-37-01

**Review by bbecquet — COMMENTED** [review]
This works great :+1:
The code should be refactored to be more React though… and it will actually make things simpler :)

File: src/panel/direction/MobileRoadMapPreview.jsx (hunk 2 above)
Comment (xem) [reply src/panel/direction/MobileRoadMapPreview.jsx:17]: Hi, thanks for the review, I actually started by doing this stuff in the proper "react" way, but when I did that, I had the impression that the snap-scroll seemed more janky, or buggy when the setState happened in the middle of a scroll. Maybe it was just a problem with my PC or my Chrome's mobile view, we could put the setStates back and verify it on real mobile devices if you want ...?

Comment (bbecquet) [reply src/panel/direction/MobileRoadMapPreview.jsx:17]: Hi!
That would be great to restore, yes :)

About the perf, there are common problems, not related to React in itself, when you plug DOM modifications directly to scroll events, as these are generated very quickly and can cause the DOM to "choke" with too many changes.

But I think if you keep you test in the scroll method to avoid doing a `this.setState({ currentStep })` when it's not needed, it should be fine. And if not we can still using throttling strategies but probably it won't be needed here.

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 10]
```diff
@@ -2,62 +2,79 @@ import React from 'react';
...
 
-  state = {
-    currentStep: 0,
-  }
+  state = { };
 
   componentDidMount() {
-    this.zoomToCurrentStep();
-    fire('move_mobile_geolocation_button', 65);
+    this.steps = document.querySelector('.mobile-roadmap-preview-steps');
```
Comment (G-Ray) [inline src/panel/direction/MobileRoadMapPreview.jsx:15]: I think we could use ref instead

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 11]
```diff
@@ -2,62 +2,80 @@ import React from 'react';
...
+  };
 
-    return <React.Fragment>
-      <div className="itinerary_mobile_step">
-        <CloseButton onClick={this.props.onClose} />
-        <RoadMapStep step={currentStep} />
-      </div>
+  toggleSize = () => {
+    this.setState( { expanded: !this.state.expanded } );
+    fire('move_mobile_bottom_ui', this.stepsRef.current.offsetHeight);
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:46]: `setState` and the triggered `render` can be async, so here you're not guaranteed that the `offsetHeight` will be the new one.
This is what you see in action here, where things are desynchronized.
Peek 08-12-2020 12-07

To solve this, we must be sure before firing the event that the component has been updated after the `setState`. This is what the `componentDidUpdate` is for, so you can use it to check what state value has changed and trigger the right events while being sure the render pass has been made.

This is one solution, but this is the "old way". I think this is the good occasion to convert this component to a functional one, when such actions (actually side-effects) can be written much more naturally:
```js
const [expanded, setExpanded] = useState(false); // functional way of declaring state
// ...
useEffect(
    () => { /* ... */  }, // do what you want…
    [expanded]  // … everytime "expanded" changes
);  
```

This would require other changes to the component, but I think it's worth it.
Would you want to try?

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 12]
```diff
@@ -1,63 +1,98 @@
...
+    // If it has changed, save it and highlight it as the current step
+    if (currentStep !== newStep) {
+      setCurrentStep(newStep);
+      fire('zoom_step', steps[newStep]);
+    }
+  };
+
+  const toggleSize = () => {
+    setExpanded(!expanded);
+    fire('move_mobile_bottom_ui', stepsRef.current.offsetHeight);
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:33]: Same here, just focus on modifiying the state and let side effects do the work for you with a state that is sure up-to-date.

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 13]
```diff
@@ -1,63 +1,98 @@
...
+  };
+
+  // On load, zoom on step 0, unexpanded
+  useEffect(
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:37]: You don't need three separate `useEffect`. Two should be enough, because you have two side effects:
 - an event firing when `expanded` value change (including on first render)
 - an event firing when `currentStep` value change (including on first render)
  
```js
useEffect(
    () => { fire('move_mobile_bottom_ui', stepsRef.current.offsetHeight); },  // with probably a check for stepsRef because it will be null on first render
    [expanded /*, ... */]
);

useEffect(
    () => { fire('zoom_step', steps[currentStep]); },
    [currentStep /*, ... */]
);
```

In functional components you don't think in terms of mount=>updated=>unmount, but in terms of which component "state" (actually a mix of its React state, its props, its refs) value triggers which side effect. 
And don't worry, it's perfectly normal to go through these steps after time spent on class components :smiley: 
It took me sooo much time to understand most of the time it's simpler like this, and I'm still not fluent.

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 14]
```diff
@@ -1,63 +1,98 @@
...
+    },
+    []
+  );
+
+  // When expanded is modified, move the map's bottom UI
+  useEffect(
+    () => {
+      fire('move_mobile_bottom_ui', stepsRef.current.offsetHeight);
+    },
+    [expanded]
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:50]: You should see linter warnings about missing stuff here. Because anything used by the side effect has to be considered as possibly changing, and then declared in the array.
Here, you'll have to add `stepsRef`, as there is no guarantee another part of the component won't trigger a change of this ref, so you want to be sure this effect will be run again in this case.

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 15]
```diff
@@ -1,63 +1,98 @@
...
+  );
+
+  // When currentStep is modified, zoom on it
+  useEffect(
+    () => {
+      fire('zoom_step', steps[currentStep]);
+    },
+    [currentStep]
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:58]: Same for `steps` here. And this is normal, if for any reason the `steps` prop changes, you'll want to trigger the effect.

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 16]
```diff
@@ -1,63 +1,98 @@
...
-        />
-      </div>
-    </React.Fragment>;
-  }
-}
+
+    // If it has changed, save it and highlight it as the current step
+    if (currentStep !== newStep) {
+      setCurrentStep(newStep);
+      fire('zoom_step', steps[newStep]);
```
Comment (bbecquet) [inline src/panel/direction/MobileRoadMapPreview.jsx:27]: This will be managed by the corresponding useEffect so it can be removed from here.

**Comment by G-Ray** [comment]
It works really well !
However I have some hiccups when scrolling until the very last item. Am I the only one who experience this behavior ?
Peek 2020-12-09 17-47

File: src/scss/includes/direction.scss [hunk 17]
```diff
@@ -395,8 +326,138 @@ button.direction_shortcut {
...
+
+    .mobile-roadmap-preview-close {
+      border: none;
+      padding: 0 0 0 8px;
+      margin: 0 0 0 11px;
+      background: $background;
+      width: 32px;
+      height: 32px;
+      border-radius: 50%;
+      box-shadow: 0 1px 4px 0 rgba(12, 12, 14, 0.2), 0 0 2px 0 rgba(12, 12, 14, 0.12);
```
Comment (G-Ray) [inline src/scss/includes/direction.scss:344]: you ca use $shadow which is defined in `colors.scss`

File: src/scss/includes/direction.scss [hunk 18]
```diff
@@ -395,8 +326,138 @@ button.direction_shortcut {
...
   .autocomplete_suggestions {
     height: 100vh;
   }
+
+  // Step by step mobile view
+  .itinerary_mobile_step_by_step {
+
+    .mobile-roadmap-preview-close {
+      border: none;
+      padding: 0 0 0 8px;
```
Comment (G-Ray) [inline src/scss/includes/direction.scss:338]: I would suggest to use our new spacing values :)

File: src/scss/includes/direction.scss [hunk 19]
```diff
@@ -395,8 +326,138 @@ button.direction_shortcut {
...
+        user-select: none;
+        flex: 0 0 calc(100vw - 70px);
+        margin: 0 6px;
+        display: flex;
+        align-items: center;
+        justify-content: center;
+        scroll-snap-align: start;
+        border: none;
+        border-radius: 12px;
+        box-shadow: 0 1px 4px 0 rgba(12, 12, 14, 0.2), 0 0 2px 0 rgba(12, 12, 14, 0.12);
```
Comment (G-Ray) [inline src/scss/includes/direction.scss:401]: $shadow

File: src/scss/includes/direction.scss [hunk 20]
```diff
@@ -395,8 +326,138 @@ button.direction_shortcut {
...
+          background-color: $grey-lighter;
+
+          .itinerary_roadmap_instruction:after {
+            background: linear-gradient(#e8e8ed77, $grey-lighter);
+          }
+        }
+
+        &.active {
+          border: 1px solid $action-blue-dark;
+          box-shadow: 0 0 0 4px rgba(128, 172, 255, 0.2);
```
Comment (G-Ray) [inline src/scss/includes/direction.scss:413]: this value is defined as`$shadow-focus` in `colors.scss`

File: src/panel/direction/MobileRoadMapPreview.jsx [hunk 21]
```diff
@@ -1,68 +1,87 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
```
Comment (G-Ray) [inline src/panel/direction/MobileRoadMapPreview.jsx:1]: :+1:

File: src/panel/direction/RoadMapItem.jsx [hunk 22]
```diff
@@ -3,11 +3,11 @@ import classnames from 'classnames';
 
 import { Divider, Flex } from 'src/components/ui';
 
-const RoadMapItem = ({ children, icon, distance, className, line, ...rest }) =>
+const RoadMapItem = ({ children, icon, distance, className, line, flex, ...rest }) =>
```
Comment (bbecquet) [inline src/panel/direction/RoadMapItem.jsx:6]: `flex` isn't a good param name, we don't understand what it does. `alignItems` would be better.
And you could give it a default value equals to `center`, that way you don't need to modify other files not concerned by this PR and manage the special step-by-step case as a kind of exception (which it is).

**Comment by xem** [comment]
> 
> 
> It works really well !
> However I have some hiccups when scrolling until the very last item. Am I the only one who experience this behavior ?
> Peek 2020-12-09 17-47

I also have random hiccups on Chrome PC + mobile simulator, but Benjamin said it was fine on real devices (to be confirmed maybe?)

**Comment by bbecquet** [comment]
> > It works really well !
> > However I have some hiccups when scrolling until the very last item. Am I the only one who experience this behavior ?
> 
> I also have random hiccups on Chrome PC + mobile simulator, but Benjamin said it was fine on real devices (to be confirmed maybe?)

@xem @G-Ray 
Well, I see the same thing of the responsive view of Chrome dekstop, but on my real Android Chrome, I don't know, the whole app has always been laggy anyways :man_shrugging:
On iOS hovewer, it behaves really well.

**Review by bbecquet — APPROVED** [review]

**Comment by amatissart** [comment]
I don't reproduce the hiccups you describe.  
But I noticed it's no longer possible to move or rotate the map when the step-by-step mode is on. The map position will stick to the position fitting the current step. Is it an expected behavior ?

**Comment by xem** [comment]
@amatissart it is not, I'm looking into it (I suspect a kind of hidden infinite loop that provokes the scroll hiccups/lags and this... I just need to find where it is :) )

**Comment by bbecquet** [comment]
> @amatissart it is not, I'm looking into it (I suspect a kind of hidden infinite loop that provokes the scroll hiccups/lags and this... I just need to find where it is :) )

@xem apparently the `zoom_step` event is fired when we touch the map, causing the view to always recenter on the current step. Let's move up the chain to understand why:
 1. according to the `useEffect` declaration, this can happen only when the `currentStep` state value or the `steps` prop changes
 2. as you took care to change `currentStep` only when needed, it's probably because the `steps` array changes when we click the map
 3. this can be confirmed by removing `steps` from the array of variables tracked for changes across render: if you remove it, it works
 4. in `DirectionPanel`, `steps` is passed as a dynamically-computed array each time we render `MobileRoadMapPreview`. It's logical that the array is considered different each time, because React will compare by ref, not content.
```jsx
<MobileRoadMapPreview
    steps={getAllSteps(activePreviewRoute)}
    onClose={this.onClose}
/>
```

So, to sum things up, we have to understand why moving the map causes a re-render of the `DirectionPanel`.
Looking quickly at it, it seems unnecessary and costly, so it may be a good target to bring optimization by removing it.
And if it proves useful at the end, we'll have to find a way to avoid recomputing the array each time.

**Comment by bbecquet** [comment]
@xem @amatissart
I put a break point in the `DirectionPanel` `render` method and found the guilty state change in the callstack:

Capture d’écran de 2020-12-15 10-20-29

Conclusion: it has nothing to do with the direction feature itself, it's related to the recent change I made to minimize the mobile Panel on any map interaction. This triggers a `setState` which will re-render any component attached to the device context (in practice, all panels I think), even if the value itself hasn't changed (that's the confusing way `setState` works).

This can be fixed by avoiding the `setState` when `panelSize` already is `minimized`, or implementing a `shouldComponentUdpate`, or using a functional component for `PanelManager` so that state updates are atomic and don't result in a re-render if no change is made…

As it's not related to the step-by-step but would be an optim for the whole app, I propose we merge this PR and I do a separate  fix.
